examples/Example17/src/PrimaryGeneratorAction.cc

```diff
@@ -47,45 +50,49 @@ void PrimaryGeneratorAction::GeneratePrimaries(G4Event *aEvent)
   {
     fHepmcAscii->GeneratePrimaryVertex(aEvent);
   }
-  else
+  else 
   {
-  G4ThreeVector oldDirection = fParticleGun->GetParticleMomentumDirection();
-  // randomize direction if requested
-  if(fRndmDirection > 0.) {
-
-    // calculate current phi and eta
-    double eta_old = atanh(oldDirection.z());
-    double phi_old = atan(oldDirection.y()/oldDirection.z());
-
-    // Generate new phi and new eta in a ranges determined by fRndmDirection parameter
-    const double phi = phi_old + (2. * M_PI * G4UniformRand()) * fRndmDirection;
-    const double eta = eta_old + (- 5. + 10. * G4UniformRand()) * fRndmDirection;
-
-    // new direction
-    G4double dirx = cos(phi) / cosh(eta);
-    G4double diry = sin(phi) / cosh(eta);
-    G4double dirz = tanh(eta);
-
-    fParticleGun->SetParticleMomentumDirection(G4ThreeVector(dirx,diry,dirz));
-  }
-  // randomize the beam, if requested.
-  if (fRndmBeam > 0.) {
-    G4ThreeVector oldPosition = fParticleGun->GetParticlePosition();
-    G4double rbeam            = 0.5 * fRndmBeam;
-    G4double x0               = oldPosition.x();
-    G4double y0               = oldPosition.y() + (2 * G4UniformRand() - 1.) * rbeam;
-    G4double z0               = oldPosition.z() + (2 * G4UniformRand() - 1.) * rbeam;
-    fParticleGun->SetParticlePosition(G4ThreeVector(x0, y0, z0));
-    fParticleGun->GeneratePrimaryVertex(aEvent);
-    fParticleGun->SetParticlePosition(oldPosition);
-  } else {
-    fParticleGun->GeneratePrimaryVertex(aEvent);
-  }
-  fParticleGun->SetParticleMomentumDirection(oldDirection);
+    G4ThreeVector oldDirection = fParticleGun->GetParticleMomentumDirection();
+    // randomize direction if requested
+    if (fRndmDirection > 0.) {
+
+      // calculate current phi and eta
+      double eta_old = atanh(oldDirection.z());
+      double phi_old = atan(oldDirection.y() / oldDirection.z());
+
+      // Generate new phi and new eta in a ranges determined by fRndmDirection parameter
+      const double phi = phi_old + (2. * M_PI * G4UniformRand()) * fRndmDirection;
+      const double eta = eta_old + (-5. + 10. * G4UniformRand()) * fRndmDirection;
+
+      // new direction
+      G4double dirx = cos(phi) / cosh(eta);
+      G4double diry = sin(phi) / cosh(eta);
+      G4double dirz = tanh(eta);
+
+      fParticleGun->SetParticleMomentumDirection(G4ThreeVector(dirx, diry, dirz));
+    }
+    // randomize the beam, if requested.
+    if (fRndmBeam > 0.) {
+      G4ThreeVector oldPosition = fParticleGun->GetParticlePosition();
+      G4double rbeam            = 0.5 * fRndmBeam;
+      G4double x0               = oldPosition.x();
+      G4double y0               = oldPosition.y() + (2 * G4UniformRand() - 1.) * rbeam;
+      G4double z0               = oldPosition.z() + (2 * G4UniformRand() - 1.) * rbeam;
+      fParticleGun->SetParticlePosition(G4ThreeVector(x0, y0, z0));
+      fParticleGun->GeneratePrimaryVertex(aEvent);
+      fParticleGun->SetParticlePosition(oldPosition);
```

Comment:
This seems to override the randomized gun position. So we need to make sure that we randomize the position for all types of gun in case ` fRndmBeam > 0`